### PR TITLE
Probable fix for out of place BinaryOpScalar bad values and/or IMAs on 11.2 (ci-all edition)

### LINKED ISSUE #51992: test_optim.py failures for CUDA 11.2 on CI
## 🐛 Bug

The following tests in test_optim.py fail in CI on CUDA 11.2 in linux and windows.
test_adadelta
test_adam
test_adamw
test_multi_tensor_optimizers
test_rmsprop

## To Reproduce

1. Run the CI jobs pytorch_linux_xenial_cuda11_2_cudnn8_py3_gcc7_test and pytorch_windows_vs2019_py36_cuda11.2_test2
2. Get the following logs with error messages:
Linux: https://app.circleci.com/pipelines/github/pytorch/pytorch/271291/workflows/ffe1be08-5d55-459e-98bd-200d6387e41c/jobs/10779915
Windows: https://app.circleci.com/pipelines/github/pytorch/pytorch/271865/workflows/2ac5bd63-6dba-4169-9f5d-7ccaf7ab33b4/jobs/10806271

## Expected behavior

Those tests should pass.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @seemethere @malfet @walterddr @pytorch/pytorch-dev-infra @ngimel

### PR DESCRIPTION
**Pull request by mcarilli**
Should close https://github.com/pytorch/pytorch/issues/51992.

ci-all resubmit of https://github.com/pytorch/pytorch/pull/52591. The plot also thickened considerably since then. Every foreach functor, it turns out, has bad `r_args` accesses for certain code paths and instantiations.

Also, I noticed the [`n % kILP == 0`](https://github.com/pytorch/pytorch/blob/2680ff7759d8a441eada383ba7aa0fa42c7d35ed/aten/src/ATen/native/cuda/ForeachFunctors.cuh#L87) condition for vectorization in all functors is way too restrictive: it'll refuse to vectorize anything on any tensor whose overall numel is not a multiple of ILP. That's out of scope though.

### CODE REVIEW

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 1]
```diff
@@ -274,7 +276,8 @@ struct ZeroFunctor {
             }
             else {
                 for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
-                    load_args<depth>(r_args, args, i_start, chunk_size, n);
+                    // Why were we loading anything when we're just going to zero it?
+                    // load_args<r_args_depth>(r_args, args, i_start, chunk_size, n);
```
Comment (izdeby) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:280]: Probably can be removed.

**Review by izdeby — APPROVED** [review]

**Comment by ngimel** [comment]
@mcarilli if you have a local repro, please add a test to test_foreach that would fail without this fix (tensor - scalar binary op with non-vectorized tail should work, looks like it's missing, other ops must be failing too).

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh [hunk 2]
```diff
@@ -316,7 +319,7 @@ struct UnaryOpFunctor {
             }
             else {
                 for(int i_start = 0; i_start < n && i_start < chunk_size; i_start += blockDim.x * kILP) {
-                    load_args<depth>(r_args, args, i_start, chunk_size, n);
+                    load_args<r_args_depth>(r_args, args, i_start, chunk_size, n);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ForeachFunctors.cuh:322]: it's always 1, right?

Comment (mcarilli) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:322]: Yes, but r_args_depth is used to declare r_args, so i wanted to be symmetric.

**Comment by mcarilli** [comment]
> if you have a local repro

I have "local repro" with just sqrt and mul, definitely not enough by itself.. I'll look at existing tests and see what isn't covered.

File: aten/src/ATen/native/cuda/ForeachFunctors.cuh (hunk 1 above)
Comment (mcarilli) [reply aten/src/ATen/native/cuda/ForeachFunctors.cuh:280]: done

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**